src/main/java/com/google/codeu/data/Datastore.java

```diff
@@ -81,4 +81,33 @@ public void storeMessage(Message message) {
 
     return messages;
   }
+
+    /** Stores the User in Datastore. */
+    public void storeUser(User user) {
```

Comment:
Consider the following scenario:

      // Request 1 (submitted by "foo@domain.tld")
      User user = ...
      storeUser(user);

      // Request 2 (also submitted by foo@domain.tld to update the "about me" data)
      User user = ....
      storeUser(user);

In this scenario, it would appear that you create duplicate entries/rows for the same user.

I would recommend, instead defining:

     public Optional<Key> getUserKeyByEmail(String email);
     public Optional<User> getUserByEmail(String email);
     public Key createOrUpdateUser(User user);

With:

     public createOrUpdateUser(User user) {
        Optional<Key> existingKeyOrAbsent = getUserKeyByEmail(user.getEmail());
        if (existingKeyOrAbsent.isPresent()) {
           Key existingKey = existingKeyOrAbsent.get();
           updateUser(existingKey, user);
           return existingKey;
         }
         return createUser(user);
     }

     private updateUser(Key key, User user) {
        // ...
     }

     private Key createUser(User user) {
        // ...
     }

Reply:
Ah, this code seems very similar to a HashMap which is going to be very good if we have high volume / want fast lookups. I believe that we can still use the current implementation and circumvent the proposed scenario if we had some checks in storeUser() to see if an user is already in the database or not. 

But, I think that it would be better to use the proposed implementation because of it allows us to separate the updating and creating the user. So later down the line, we can implement different features and attributes for the users if we wanted to.

Reply:
Re: your feedback, @markser 

The observation that the Cloud Datastore is similar to HashMap is a good one. It does a little more than a HashMap in that it also supports indexing by multiple different entity properties (not just the key), but it is definitely similar in that each entity (row) is assigned a unique key that identifies it and that it is quick to find an entity by key (and less quick to query rows that meet some criteria).

In terms of making look-ups quick, there is a trade-off to be aware of here between consistency and speed. When storing data in a cache, there is a possibility of serving an old version from the cache after the data has been updated in the authoritative database. That issue can be mitigated with cache invalidation, but such invalidation requires centralization of the caching (rather than having each copy of the program cache in its own process's virtual memory), and that centralization, in turn, trades off some speed and availability for that consistency.

In typical real-world applications that need to speed up such lookups, you'll find that they often use a caching server (like Redis or Memcache) in conjunction with the database for caching with invalidation, expiration.

As with most optimizations, though, unless the optimization is non-trivial, such optimizations are best to defer unless/until the need for the optimization can be clearly rationalized relative to other priorities. At this point, adding Redis or Memcache to the system strikes me as premature.